ENH: Make it possible to NpyIter_RemoveAxis an empty dimension

seberg opened the pull request:
This is the followup to that "kludge" discussion. I did not really check much if there might be something going awefully wrong, though on the other hand it is hard to imagine something too bad.
## 

There are certainly more kludges to remove here assuming that it works. So should check for those first.

charris commented:
@seberg I've forgotten, if I ever knew, what this is about. Seems like a pretty small patch and if it fixes other problems I'd like it to go in.

seberg commented:
Its about gufuncs not being able to handle empty arrays. unfortunatly, lapack doesn't handle, so that module needs to be carefully checked ebfore putting this in.

seberg commented:
OK, I have updated this, so some of the things now accept 0-d such as determinant. Cholesky I am not sure, etc. should have a bit careful look over the test case and everything that was changed has a test case etc. (i.e. I am a bit worried about the linalg changes, less so the nditer ones).
I can stash things later, and it needs a release notes comment as well, since nditer is changed (and also linalg).

homu commented:
:umbrella: The latest upstream changes (presumably #6115) made this pull request unmergeable. Please resolve the merge conflicts.

charris commented:
@seberg Could you rebase this? I'm going to branch 1.10 tomorrow, and if you think this is ready I'll put it in.

seberg commented:
Probably is pretty much, but I am on vacation and do not have my compiter with me. But maybe it is better to wait for bew master anyway, in case I missed a function.

seberg commented:
Rebased this, but if someone has the patience to double check the linalg stuff, maybe would not hurt.

homu commented:
:umbrella: The latest upstream changes (presumably #6905) made this pull request unmergeable. Please resolve the merge conflicts.

homu commented:
:umbrella: The latest upstream changes (presumably #7073) made this pull request unmergeable. Please resolve the merge conflicts.

seberg commented on numpy/linalg/linalg.py:
```diff
@@ -1010,8 +991,6 @@ def eigvalsh(a, UPLO='L'):
     _assertRankAtLeast2(a)
     _assertNdSquareness(a)
     t, result_t = _commonType(a)
-    if _isEmpty2d(a):
```
There was a bug here  for complex value, since it missed the cast to real (I think in one other function as well).

seberg replied:
@eric-wieser can you have a look if you have time. Since I don't know all your newer changes, and I guess it may be that the tests are not quite how they would be nice (dunno how well the rebase worked). Also this moves some of your fixups into the C-lapack wrappers.

eric-wieser commented on numpy/linalg/linalg.py:
```diff
@@ -367,21 +367,8 @@ def solve(a, b):
     else:
-        if b.size == 0:
-            if (a.shape[-1] == 0 and b.shape[-2] == 0) or b.shape[-1] == 0:
-                a = a[:,:1].reshape(a.shape[:-1] + (1,))
```
This looks to me like it was wrong in the first place, and should have been `a[...,:1]`

eric-wieser commented on numpy/linalg/umath_linalg.c.src:
```diff
@@ -1092,7 +1092,10 @@ static inline void
-    LAPACK(@cblas_type@getrf)(&m, &m, (void *)src, &m, pivots, &info);
+    if (m != 0) {
+        /* Lapack (at least some versions) do not handle n == 0 */
+        LAPACK(@cblas_type@getrf)(&m, &m, (void *)src, &m, pivots, &info);
```
Which versions are we supporting? Even the oldest 3.0.0 version in lapack_lite we're using supports this, but the third m should be `max(1,M)` according to the documentation

seberg replied:
Dunno, my guess was I had problems on my system, or, I just did not know if it was necessary or not. The `max(1, m)` thing may just be the reason too I suppose.

seberg replied:
@eric-wieser is this the case for all of these functions? Or which one would need this kind of max handling? I will add a commit undoing it and see what happens for now, and we can still go back and forth if problems show up with some lapack or so.

seberg replied:
OK, it worked here, but for none of the other cases, but I guess you expected that?

seberg replied:
Though, I did not try to replace parameters with 1 instead of 0 there. Is it worth checking that out?

eric-wieser replied:
You'd need to look at the documentation for each lapacj function in turn, which can be found within f2c_[czds]_lapack.c

seberg replied:
Hmmm, ok, should do some time I guess, but don't have the nerve for it right now.

eric-wieser reviewed (commented):
Looks generally good, but a couple of comments:

* A lot of lapack functions do support `n == 0`, but have an argument that we currently pass as `n` that we should be passing as `max(n, 1)` according to the docs
* The test-suite already tests some empty arrays through the `do` method, so maybe the test_size_0 could be merged, if we really care (not sure we do). At any rate, I think that adding some dtype-changing tests to the main set of tests would be sensible. Perhaps for another PR though

eric-wieser commented:
Do `det` and `slogdet` need their special casing still?

eric-wieser commented:
Might be tempted to merge this without any lapack changes at all and save that for a future pr

seberg commented:
True, though it does fix a bug somewhere with the result type which we have to fix otherwise then.

eric-wieser commented:
Does that one not now fall freely out of the gufunc if you move the size check to there?

eric-wieser commented:
So what I'm proposing, is remove all the python checks, and just keep the ` /* Lapack (at least some versions) do not handle n == 0 */` guard clauses that you already have in place (but with a comment that emphasizes that more work is needed, not that lapack can't support it).

That way, if it turns out that some lapack-emulations do fail with zeros sizes, we have something to roll back onto.

seberg commented:
Ah OK, yeah that is an option of course, and I don't think I mind it. But if we do it, we have to be prepared for it sticking around for the next 5 years....

eric-wieser commented:
Right, but that's better than that _and_ this sticking around for 5 years!

eric-wieser commented:
Regarding fixing lapack, a regex for `&(\w+),[^;]*?&\1` in `umath_linalg.c.src` probably finds all the places where we're not handling this correctly

eric-wieser commented:
If you tick the "allow maintainers to edit" box, I can see if I can patch up some of the other functions

seberg commented:
Ah, very old PR, hehe. Done.

eric-wieser commented:
@seberg: Ok, fixed another one. This one raised why we want #8381

eric-wieser commented:
Ok, that should be the lot of them. Pretty much everything I added in #8368 is gutted and replaced here!

eric-wieser commented:
Bonus: svd now works too.

All that's left are `qr` and `lstsq`, which should probably be fixed by vectorizing them at the same time.

eric-wieser commented on numpy/linalg/linalg.py:
```diff
@@ -1672,19 +1643,17 @@ def pinv(a, rcond=1e-15 ):
     u, s, vt = svd(a, 0)
     m = u.shape[0]
     n = vt.shape[1]
-    cutoff = rcond*maximum.reduce(s)
```
`maximum.reduce` fails on empty arrays with `ValueError: zero-size array to reduction operation maximum which has no identity`. Shouldn't this just return `-inf` for arrays of dtype `inexact`?

seberg commented:
You have been busy. Hmmm, I am not sure about the condition number. I guess one would the only sensible choice and make sense in (almost?) all cases where one would use the condition number since the lstsq problem, etc. have an exact solution? I am not 100% sure, though. If we are 90% sure, we should probably poll the list one more time just in case someone complains.

seberg commented:
I am not sure I like the amount of valgrind output (may all be OK though, did not check too carefully) or the result including `6.92461009e-310` as a numerical value for the empty (not 0x0) cases of SVD.

eric-wieser commented:
> for the empty (not 0x0) cases of SVD.

Do you mean 0d? What was the scalar value?

eric-wieser commented:
Matlab seems to return zero for the condition number

seberg commented:
I meant non-square ones, such as `np.array([[]])` or transposed. Oh, zero seemed sensible when looking at the norms/definition on wikipedia, but I thought since 1 means "good". But... larger then one is bad, so I guess zero, being below one, also means "good", so should be the right value....

`R` may be another data point, but I think R may not have the best habit about empty arrays probably.

eric-wieser commented:
Shouldn't the svd be empty in that case?

eric-wieser commented:
> I meant non-square ones, such as np.array([[]]) or transposed.

eric-wieser commented:
Hmm, I see what you mean. Based on doing this repeated times, it's clear that the memory inside the array is uninitialized

eric-wieser commented:
What do you think the best way to move forward here is? Revert the SVD commit, squash a little, and merge?

seberg commented:
Sounds fine to me, if you have a good idea about the SVD would also be fine. I think I will also run the tests through valgrind again before the end, just to be sure. I did not think about reorganizing the tests at all, and since you have worked on these tests recently, let do whatever you think is right.

eric-wieser commented:
@seberg: I moved my above fixes to a new PR, since the bad LDA params are a bug by themselves

homu commented:
:umbrella: The latest upstream changes (presumably #8756) made this pull request unmergeable. Please resolve the merge conflicts.

eric-wieser commented:
@seberg: Rebased

seberg committed:
ENH: Allow removal of 0-sized axes from nditer and empty gufunc calls

The caller should take of not using pointers given in such a case

seberg committed:
MAINT: Remove python side empty array handling from linalg

The necessary fixup on the C-side of linalg has been done already
(i.e. the gufuncs correctly work for these empty arrays).
This also enables cholesky decomposition and fixes a small bug in pinv
handling.

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

seberg commented:
@eric-wieser thanks a lot for all the nice work, your other cleanup really made this one pretty lean indeed. I rebased/commited it down to two commits (one for the C-side and one for the linalg/python side). Should probably be good to go, but go ahead and complain if you find something :)

eric-wieser commented:
> Hmm, just noticed, you had somehow two authors, or author+commiter. Not sure how to do that, though it would seem right to add you for the second commit.

What you're seeing is author+committer. In this case, you're definitely the latter, and to be honest - mostly the former - I'm not sure I can point to a single line in the second commit that was my doing.

Arguably, that second commit is almost entirely reverting lines that I added in #8368!

seberg commented:
Hahaha, yeah, probably most of what you did ended up in the linalg fixup. Added a co-author note thingy in the commit message, if you like will change it to anything else, but I am happy with things.

eric-wieser commented:
Yep, saw that - if you think that should be there, then that's fine with me.

eric-wieser reviewed (approved):
All looks good to me, but probably needs another set of eyes since I've looked at this too much

seberg committed:
BUG: Fix matrix multiply test gufunc for empty matrices

The case where the result is not empty was missing the zeroing out
now that it is legal to use that input.

seberg commented:
Actually, `numpy.core.umath_tests.matrix_multiply` choked on the gufunc change, added a commit to fix that.

eric-wieser commented:
To be clear, that `gufunc` is not for public use, and just for testing?

seberg commented:
Sure, fixed it anyway (though there are probably nicer ways to fix it).

eric-wieser commented on numpy/core/src/umath/umath_tests.c.src:
```diff
@@ -155,6 +155,21 @@ static void
+                    op  +=  os_p;
+                }
+                op  -=  ob_p;
+            op  +=  os_m;
```
indentation

eric-wieser commented on numpy/core/src/umath/umath_tests.c.src:
```diff
@@ -155,6 +155,21 @@ static void
+            }
+        END_OUTER_LOOP
+        return;
+    }
```
Wouldn't a better fix be to just reassign the pointers, and run the below code?

Something like

```python
if (dn == 0) {
     @typ@ ip_val = 0;
     @typ@ ip2_val = 0;
      ip1 = ip2 = &ip_val;
      // set strides to 0
      dn = 1
}
```

seberg replied:
Nice hack, didn't think about it much.

seberg replied:
Hmm, no, the `char *ip1=args[0], *ip2=args[1], *op=args[2];` line is important where it is, since the macro will manipulate `args`

eric-wieser commented:
> though there are probably nicer ways to fix it).

> Would not be smart to use it anyway

Never mind then.

eric-wieser commented:
Yeah, with the innermost loop over n, this just falls out:
```C
BEGIN_OUTER_LOOP_3
    char *ip1=args[0], *ip2=args[1], *op=args[2];
    for (m = 0; m < dm; m++) {
        for (p = 0; p < dp; p++) {
            *(@typ@ *)op = 0;
            for (n = 0; n < dn; n++) {
                @typ@ val1 = (*(@typ@ *)ip1);
                @typ@ val2 = (*(@typ@ *)ip2);
                *(@typ@ *)op += val1 * val2;
                ip1 += is1_n;
                ip2 += is2_n;
            }
            ip1 -= ib1_n;
            ip2 -= ib2_n;
            ip2 += is2_p;
            op  +=  os_p;
        }
        ip2 -= ib2_p;
        op  -=  ob_p;
        ip1 += is1_m;
        op  +=  os_m;
    }
END_OUTER_LOOP
```

homu commented:
:umbrella: The latest upstream changes (presumably #8847) made this pull request unmergeable. Please resolve the merge conflicts.

eric-wieser commented:
This didn't previously segfault, did it? Bad rebase?

eric-wieser commented:
One failure is:

```
Fatal Python error: GC object already tracked
Current thread 0x00007fb6e0f0e700 (most recent call first):
  File "/home/travis/build/numpy/numpy/venv-for-wheel/lib/python3.5/site-packages/numpy/core/tests/test_ufunc.py", line 694 in test_object_scalar_multiply
```

eric-wieser commented on doc/release/1.13.0-notes.rst:
```diff
@@ -120,6 +140,8 @@ behaviour is recovered if ``axis=None`` (default).
 ------------------------------------------------
 Users can now specify a not-constant spacing for data. 
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Users can now specify a not-constant spacing for data.
```
Bad merge

eric-wieser commented on doc/release/1.13.0-notes.rst:
```diff
@@ -131,11 +153,13 @@ In particular ``np.gradient`` can now take:
 
 This means that, e.g., it is now possible to do the following::
     
+This means that, e.g., it is now possible to do the following:
```
Bad merge

eric-wieser commented on doc/release/1.13.0-notes.rst:
```diff
@@ -227,6 +251,8 @@ More ``linalg`` operations now accept empty vectors and matrices
 arrays with a 0 in the last two dimensions: ``det``, ``slogdet``, ``pinv``,
 ``eigvals``, ``eigvalsh``, ``eig``, ``eigh``.
+arrays with a 0 in the last two dimensions: `det``, ``slogdet``, ``pinv``,
+``eigvals``, ``eigvalsh``, ``eig``, ``eigh``, ``cholesky``.
```
Bad merge

seberg replied:
Ooops, I am sorry, I am a bit overwhelmed with other things right now, not sure I will be checking on things regularly for a while :/.

seberg replied:
OK, that should have fixed it up, when going back to the previous version of setting to zero, I got the pointer arithmetic wrong apparently.

seberg replied:
ok, now for real... ;)

homu commented:
:umbrella: The latest upstream changes (presumably #8999) made this pull request unmergeable. Please resolve the merge conflicts.

eric-wieser commented:
Nice job! Based on the [silence on the waiting list](http://numpy-discussion.10968.n7.nabble.com/heads-up-gufuncs-on-empty-arrays-and-NpyIter-removal-of-empty-axis-td44118.html), this looks pretty safe to merge?

charris commented:
I don't have a problem with this getting merged. Support for empty arrays is not complete and this is another step on the way,

seberg commented:
OK; lets give it a shot.

eric-wieser commented:
On the subject of empty array kludges - what is going on [here](https://github.com/seberg/numpy/blob/1c4e0d4d1b8e365111b70c559e6b66dd34c35d06/numpy/core/src/umath/ufunc_object.c#L2481-L2516)? I can't see how the iteration can be empty, yet the results not be

eric-wieser commented:
Apparently not a single one of our tests hits that code

seberg commented:
Sounds like it was planned to catch those empty array matrix multiplies for us, but the if is incorrect. Since this is not really a reduction.... I would tend to leave it as is after this now (i.e. let the inner loop worry about it) and remove the code block (assuming, that we are not missing something really silly).

eric-wieser commented:
Doesn't look like it's checking for the right thing - that's looking at the `...` in `(..., m, n) -> (..., n, o)`, right? So if `...` is empty, the output is always empty?

seberg commented:
Yeah.
